pkg/util/modules/modules.go

```diff
@@ -0,0 +1,157 @@
+package modules
+
+import (
+	"fmt"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
+	"github.com/pkg/errors"
+)
+
+type service func() (services.Service, error)
+
+// module is the basic building block of the application
+type module struct {
+	// dependencies of this module
+	deps []string
+
+	// initFn for this module (can return nil)
+	initFn service
+}
+
+// Manager is a component that initialises modules of the application
+// in the right order of dependencies.
+type Manager struct {
+	modules map[string]module
+}
+
+// NewManager creates a new Manager
+func NewManager() *Manager {
+	return &Manager{
+		modules: make(map[string]module),
+	}
+}
+
+// RegisterModule registers a new module with ModuleManager
```

Comment:
```suggestion
// RegisterModule registers a new module with name and init function.
```

Can init function be nil? Must name be unique or not? Please specify.